core/sql/generator/GenRelScan.cpp

```diff
@@ -1131,16 +1139,29 @@ short FileScan::codeGenForHive(Generator * generator)
   // Try to get enough buffer space to hold twice as many records
   // as the up queue.
   //
-  // This should be more sophisticate than this, and should maybe be done
-  // within the buffer class, but for now this will do.
-  // 
+  
+  UInt32 FiveM = 5*1024*1024;
+
+  // If returnedrowlen > 5M, then set upqueue entries to 2.
+  if (returnedRowlen > FiveM)
+    upqueuelength = 2;
+
   ULng32 cbuffersize = 
     SqlBufferNeededSize((upqueuelength * 2 / numBuffers),
 			returnedRowlen);
   // But use at least the default buffer size.
   //
   buffersize = buffersize > cbuffersize ? buffersize : cbuffersize;
 
+  // Cap the buffer size at 5M and adjust upqueue entries.
+  // Do this only if returnrowlen is not > 5M
+  if ((returnedRowlen <= FiveM) && (buffersize > FiveM))
```

Comment:
What happens if the returned row length is 3Mb. Do we waste 2Mb in buffer space?